### ISSUE TITLE
Consider adding operators to FFI `Pointer` objects and deprecate/remove not-very-idiomatic `elementAt`

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#54250]
If we have a `Pointer<Uint8>` and would like to get one that is 1 bytes further we can
```dart
Pointer<Uint8> p;

final p1 = p.elementAt(1);
final p1 = Pointer.fromAddress(p.address + 1);
```
None of which seem very nice.

Also the name `elementAt` would suggest one actually gets the element at a certain offset, which it doesn't, it gets the address of the element at an offset.

In C one would simply
```c
uint8_t* p = ...;

uint8_t* p10 = p + 1;
```

We should consider adding extension methods such as
```dart
extension on Pointer<Uint8> {
  Pointer<Uint8> operator +(int offset) =>
      Pointer.fromAddress(address + offset);
}
```
then the example will turn into
```dart
Pointer<Uint8> p;
final p10 = p + 1;
```

Even the following would work out of the box:
```dart
Pointer<Uint8> p;
p++;
```

If we do so, there may no longer be a need to have `elementAt` and could deprecate & remove it.

There's a question whether we should support `operator-` and if so, whether it should take two pointers and return the difference as an integer or it should take an int and return a pointer minus the offset.

/cc @dcharkes opinions?

**Comment by dcharkes** [comment]
I am in favor of renaming `elementAt` to `operator +`.

Maybe obvious, but not mentioned above, the `address` increments would be in terms of `sizeOf<element type>()`. (To make it consistent with pointer arithmetic in C.)

> We should consider adding extension methods such as

For AbiSpecificIntegers and compounds we need some CFE magic, as we currently do for `elementAt` as well.

Currently, the extension methods for the types with known sizes are in the patch file and invisible for users, we could move these to be in the non-patchfile.

> There's a question whether we should support `operator-` and if so, whether it should take two pointers and return the difference as an integer or it should take an int and return a pointer minus the offset.

In these kind of situations it's really painful that we don't have overloading in Dart, then we could just have both. Maybe Dart should get some kind of overloading support for _static_ methods that cannot be called dynamically. https://github.com/dart-lang/language/issues/1122 cc @lrhn

**Comment by lrhn** [comment]
What we did with `DateTime` is to let `-` match `+` and take a duration, and have a separate `difference` method to give the difference between two `DateTime`s.

That is, the same strategy as ever time: different names instead of overloading.

I'd do the same here, so `+=` and `++` are matched by `-=` and `--`.
(If you have multiple subtraction operations, use `-` for the one that returns the same type as the receiver, so the composite operations work.)

**Comment by shikharish** [comment]
Hello, I would like to work on this issue.

**Comment by shikharish** [comment]
Hello, can someone help me understand how the extension method should be implemented.
![image](https://github.com/dart-lang/sdk/assets/139053348/5baca28c-4578-45f9-852d-9e662698347b)
*from ffi.dart file*

This is how I think it should be implemented(for `+`).

**Comment by shikharish** [comment]
@dcharkes @mkustermann 
Any guidance please?

**Comment by dcharkes** [comment]
> Any guidance please?

One way to find the relevant places to update is to follow the git blame trail on `elementAt`:

* https://dart-review.googlesource.com/c/sdk/+/276102

Most of these files will have to be edited:

* sdk/lib/ffi/ffi.dart 
  * The public API
* runtime/tools/ffi/sdk_lib_ffi_generator.dart 
  * The code in the public API is generated by this file.
* sdk/lib/_internal/vm/lib/ffi_patch.dart 
  * The implementation for fixed size integers is generated by the same file into this file.
* pkg/vm/lib/transformations/ffi/use_sites.dart 
  * Contains the implementation for AbiSpecificInteger elementAt, struct elementAt and union elementAt.
* pkg/vm/lib/transformations/ffi/common.dart  
  * Contains some imports for the above file. 

Also, there are tests in

* tests/ffi/
  *  search for `elementAt(`. All of these need to be duplicated for `+`, `-`, `+=`, and `-=`.

You will need to build the SDK locally to test all of this.
Please take a look at https://github.com/dart-lang/sdk/blob/main/CONTRIBUTING.md

**Comment by shikharish** [comment]
Thank you very much.
I have set up my local development environment and tested for `+` operator(not rigorously though yet). I will make all the necessary changes in the files.

Regarding `-` operator,

> I'd do the same here, so `+=` and `++` are matched by `-=` and `--`. (If you have multiple subtraction operations, use `-` for the one that returns the same type as the receiver, so the composite operations work.)

It is not clear to me how `+` would be *matched by*  `-`. Can you please elaborate or share some readings related to this?

**Comment by dcharkes** [comment]
> It is not clear to me how `+` would be _matched by_ `-`. Can you please elaborate or share some readings related to this?

I was referring to:

> What we did with `DateTime` is to let `-` match `+` and take a duration, and have a separate `difference` method to give the difference between two `DateTime`s.
> 
> That is, the same strategy as every other time: different names instead of overloading.
> 
> I'd do the same here, so `+=` and `++` are matched by `-=` and `--`. (If you have multiple subtraction operations, use `-` for the one that returns the same type as the receiver, so the composite operations work.)

So `pointer += -4` would be the same as `pointer -= 4` and `pointer + -4` the same as `pointer - 4`.

**Comment by dcharkes** [comment]
> But how would that be implemented? Because for operator `-`, if we do `ptr1 - 4` then it would return a pointer but `ptr1 - ptr2` would return an int.

No, `ptr1 - ptr2` would not be supported the operand to minus would be an int `Pointer operator -(int value)`.

You could add a `difference` method as well which takes another `Pointer` and returns an int: `int difference(Pointer p) => p.address - this.address;`. Then it would be consistent with `Duration`/`DateTime` as Lasse suggested. But I'm fine with leaving `difference` out.

**Comment by shikharish** [comment]
Alright, thank you. Will add the operators first.

**Comment by shikharish** [comment]
@dcharkes Most tests written use `final` Pointer variables and use of `elementAt()` does not change the original pointer. I have duplicated tests for `+` and `-` wherever possible but could not for `+=` and `-=`. 
Should I write tests for `+=`, `-=`? If yes, are there any rules or things to keep in mind for writing these tests?

**Comment by dcharkes** [comment]
> @dcharkes Most tests written use `final` Pointer variables and use of `elementAt()` does not change the original pointer. I have duplicated tests for `+` and `-` wherever possible but could not for `+=` and `-=`. Should I write tests for `+=`, `-=`? If yes, are there any rules or things to keep in mind for writing these tests?

`a += b` is basically always `a = a + b`. Which is not FFI specific, it's just a general Dart language feature. So a single new test that uses `+=` and `-=` on pointers should be fine. No need to duplicate everything. 👍

**Comment by shikharish** [comment]
@dcharkes Uploaded the patch: https://dart-review.googlesource.com/c/sdk/+/341420

**Comment by dcharkes** [comment]
Awesome @shikharish 🚀 !

I have left a round of comments there.

**Comment by dcharkes** [comment]
FYI @osa1 @eyebrowsoffire for dart2wasm. I don't believe anything needs to happen on your side. The dart2wasm implementation is reusing the use_sites.dart transformer in which this is implemented.

**Comment by shikharish** [comment]
@dcharkes I think due to a recent commit, the build command is failing so I am not able to build and test.
Opened an issue : #54321

**Comment by shikharish** [comment]
@dcharkes Issue solved and uploaded Patch 2.

**Comment by dcharkes** [comment]
> @dcharkes Issue solved and uploaded Patch 2.

👌 

@shikharish are you able to log in on https://dart-review.googlesource.com/c/sdk/+/341420 and reply to the comments there? You can mark them resolved once addressed and then click send. I'll get a notification to review the CL again that way.

**Comment by shikharish** [comment]
Yes I am logged in. I have marked some comments as "Done". I do not see a Send option though.

**Comment by dcharkes** [comment]
> Yes I am logged in. I have marked some comments as "Done". I do not see a Send option though.

It's called "Reply"

**Comment by shikharish** [comment]
This was my first contribution to Dart(and the first to such a large codebase). Hope to make more in the future! :)

**Comment by dcharkes** [comment]
Thanks for your contribution @shikharish! 🚀

**Comment by dcharkes** [comment]
@shikharish, we have branched for Dart 3.3, so we can now remove `elementAt`. Would you like to create a PR?

**Comment by shikharish** [comment]
@dcharkes Sure!

**Comment by shikharish** [comment]
@dcharkes I believe I would have to remove(or replace with operators) every instance of `elementAt` from all files in the repository? A quick search shows around 250 instances in 55 files.

**Comment by dcharkes** [comment]
> @dcharkes I believe I would have to remove(or replace with operators) every instance of `elementAt` from all files in the repository? A quick search shows around 250 instances in 55 files.

Correct. 👍  All the ones we duplicated we can remove. The ones we didn't duplicate we can replace.

**Comment by shikharish** [comment]
@dcharkes I have removed the definitions of `elementAt` from all Pointer types but I am getting a few errors while building the sdk.

```
 ../../pkg/mmap/lib/src/mmap_impl.dart:143:25: Error: The operator '+' isn't defined for the class 'Pointer<Uint8>'.
 - 'Pointer' is from 'dart:ffi'.
 - 'Uint8' is from 'dart:ffi'.
Try correcting the operator to an existing operator, or defining a '+' operator.
  return (finalizerStub + offset).cast<Void>();
                        ^
../../third_party/pkg/ffi/lib/src/utf16.dart:59:31: Error: The operator '+' isn't defined for the class 'Pointer<Uint16>'.
 - 'Pointer' is from 'dart:ffi'.
 - 'Uint16' is from 'dart:ffi'.
Try correcting the operator to an existing operator, or defining a '+' operator.
      final char = (codeUnits + i).value;
                              ^
```

These two files give an error. I do not understand why because the `+` operator is defined in `ffi.dart`. Can you pls help?

**Comment by shikharish** [comment]
@dcharkes Any ideas?

**Comment by dcharkes** [comment]
I believe these packages might be used with the checked in SDK in `tools/sdks/dart-sdk`.

@athomas could we roll the checked in SDK?

**Comment by dcharkes** [comment]
(reopening issue to track removing `elementAt`)

**Comment by athomas** [comment]
> I believe these packages might be used with the checked in SDK in `tools/sdks/dart-sdk`.
> 
> @athomas could we roll the checked in SDK?

Yes, you just need to change the DEPS to point them at a released SDK that has the necessary change.

**Comment by dcharkes** [comment]
I'm trying to get `package:ffi` updated in the Dart SDK, https://dart-review.googlesource.com/c/sdk/+/346761.

However, that is currently blocked on flutter_tools being compatible with another package in the same repo, so I'm trying to roll that: https://github.com/flutter/flutter/pull/141864.

Sorry for the delay. Breaking changes always need some "rolls" from one place to the other. And when multiple packages in a monorepo have breaking changes it gets worse. 🙈 

I'll ping the CL once it's ready to land!

**Comment by shikharish** [comment]
Wow! This was my first encounter with such a large codebase so I did not know there can be so many side-effects. I would love to work more on this. Please do ping me if there's anything :)

**Comment by dcharkes** [comment]
Unfortunately we have another use that blocks rolling this into g3:

* https://github.com/dart-windows/win32/issues/815

We need to wait with relanding until that is addressed. (We might have been a little bit too eager with the cleanup. Let's try again after Dart 3.3 is actually released and win32 has had time to migrate.)

**Comment by dcharkes** [comment]
@oprypin brought up a good point: A `pubspec.yaml` can declare an older language version. But if we remove the method, a newer Dart SDK will also not have the method anymore for the older language version.

I don't believe we have a way of deleting method only for a newer language version.

@eernstg @lrhn What's our workflow for deleting old things? If we keep things deprecated but not deleted for newer language versions, they are technically supported. So when the time comes when we bump the lowest-supported language version in the SDK all versions that have been released in the meantime technically still supported the deprecated method.

If we don't have a way of deleting methods only for newer language versions, I believe we can close this issue and just leave the deprecated method working for the foreseeable future.

**Comment by lrhn** [comment]
We don't have versioned libraries, so any solution will be ad hoc.

We can make it an error to call or access a particular function in a library above a particular language version. We did that with the `List` constructor for Dart 2.12.

For static members, it's pretty effective.
For instance members, we have to be more careful about ensuring there are no false positives, and we can't prevent dynamic invocation.

It will require some work, at least for the front ends (CFE and analyzer), so unless it's important, it's probably easier to just deprecate and remove as a breaking change "later", when most code has been migrated off it.

**Comment by dcharkes** [comment]
> so unless it's important, it's probably easier to just deprecate and remove as a breaking change "later", when most code has been migrated off it.

Okay, lets just postpone it then.

@shikharish Thanks for all the work! Sorry that we cannot land the delete CL right now. Let's come back in a year or so to try to delete it.
